In [1]:
import requests
openweathermap_url = "http://api.openweathermap.org/data/2.5/weather"

parameters = {
    'q'     : 'New York, NY, USA',
    'units' : 'imperial',
    'mode'  : 'json',
    'appid' : ' ' # Must have own appid
}
resp = requests.get(openweathermap_url, params=parameters)
data = resp.json()
data

{'coord': {'lon': -73.99, 'lat': 40.73},
 'weather': [{'id': 501,
   'main': 'Rain',
   'description': 'moderate rain',
   'icon': '10d'}],
 'base': 'stations',
 'main': {'temp': 43.86,
  'feels_like': 37.51,
  'temp_min': 41,
  'temp_max': 46.4,
  'pressure': 1020,
  'humidity': 93},
 'wind': {'speed': 8.05, 'deg': 240},
 'rain': {'1h': 1.27},
 'clouds': {'all': 90},
 'dt': 1581364783,
 'sys': {'type': 1,
  'id': 4610,
  'country': 'US',
  'sunrise': 1581335815,
  'sunset': 1581373415},
 'timezone': -18000,
 'id': 5128581,
 'name': 'New York',
 'cod': 200}

In [2]:
print('The current temp is ' + str(data['main']['temp']))

The current temp is 43.86


In [3]:
print('Weather description: ' + str(data['weather'][0]['description']))

Weather description: moderate rain


In [4]:
parameters = {
    'q'     : 'New York, NY, USA',
    'units' : 'metric',
    'mode'  : 'json',
    'appid' : ' ' # Must have own appid
}
resp = requests.get(openweathermap_url, params=parameters)
data = resp.json()
data

{'coord': {'lon': -73.99, 'lat': 40.73},
 'weather': [{'id': 501,
   'main': 'Rain',
   'description': 'moderate rain',
   'icon': '10d'}],
 'base': 'stations',
 'main': {'temp': 6.59,
  'feels_like': 3.06,
  'temp_min': 5,
  'temp_max': 8,
  'pressure': 1020,
  'humidity': 93},
 'wind': {'speed': 3.6, 'deg': 240},
 'rain': {'1h': 1.27},
 'clouds': {'all': 90},
 'dt': 1581364856,
 'sys': {'type': 1,
  'id': 4610,
  'country': 'US',
  'sunrise': 1581335815,
  'sunset': 1581373415},
 'timezone': -18000,
 'id': 5128581,
 'name': 'New York',
 'cod': 200}

In [5]:
parameters = {
    'q'     : 'San Francisco, CA, USA',
    'units' : 'metric',
    'mode'  : 'json',
    'appid' : ' ' # Must have own appid
}
resp = requests.get(openweathermap_url, params=parameters)
data = resp.json()
data

{'coord': {'lon': -76.88, 'lat': 1.17},
 'weather': [{'id': 500,
   'main': 'Rain',
   'description': 'light rain',
   'icon': '10d'}],
 'base': 'model',
 'main': {'temp': 14.21,
  'feels_like': 13.36,
  'temp_min': 14.21,
  'temp_max': 14.21,
  'pressure': 1010,
  'humidity': 85,
  'sea_level': 1010,
  'grnd_level': 754},
 'wind': {'speed': 1.98, 'deg': 147},
 'rain': {'3h': 0.38},
 'clouds': {'all': 99},
 'dt': 1581365016,
 'sys': {'country': 'CO', 'sunrise': 1581333575, 'sunset': 1581377042},
 'timezone': -18000,
 'id': 3669881,
 'name': 'San Francisco',
 'cod': 200}

In [6]:
parameters = {
    'lat'     : '1.17',
    'lon'     : '-76.88',
    'units' : 'metric',
    'mode'  : 'json',
    'appid' : ' ' # Must have own appid
}
resp = requests.get(openweathermap_url, params=parameters)
data = resp.json()
data

{'coord': {'lon': -76.88, 'lat': 1.17},
 'weather': [{'id': 500,
   'main': 'Rain',
   'description': 'light rain',
   'icon': '10d'}],
 'base': 'model',
 'main': {'temp': 14.21,
  'feels_like': 13.36,
  'temp_min': 14.21,
  'temp_max': 14.21,
  'pressure': 1010,
  'humidity': 85,
  'sea_level': 1010,
  'grnd_level': 754},
 'wind': {'speed': 1.98, 'deg': 147},
 'rain': {'3h': 0.38},
 'clouds': {'all': 99},
 'dt': 1581365017,
 'sys': {'country': 'CO', 'sunrise': 1581333575, 'sunset': 1581377042},
 'timezone': -18000,
 'id': 3669881,
 'name': 'San Francisco',
 'cod': 200}

In [7]:
url = 'http://api.ipstack.com/check?access_key=294a3b9e1df7ba2fc86e40274859521a'
resp = requests.get(url)
data = resp.json()

parameters = {
    'lat'     : data["latitude"],
    'lon'     : data["longitude"],
    'units' : 'metric',
    'mode'  : 'json',
    'appid' : ' ' # Must have own appid
}
resp = requests.get(openweathermap_url, params=parameters)
data = resp.json()
data

{'coord': {'lon': -122.08, 'lat': 37.42},
 'weather': [{'id': 800,
   'main': 'Clear',
   'description': 'clear sky',
   'icon': '01d'}],
 'base': 'stations',
 'main': {'temp': 18.99,
  'feels_like': 14.5,
  'temp_min': 15.56,
  'temp_max': 21.67,
  'pressure': 1018,
  'humidity': 17},
 'visibility': 16093,
 'wind': {'speed': 2.45, 'deg': 14},
 'clouds': {'all': 1},
 'dt': 1581365018,
 'sys': {'type': 1,
  'id': 4322,
  'country': 'US',
  'sunrise': 1581347015,
  'sunset': 1581385299},
 'timezone': -28800,
 'id': 5375480,
 'name': 'Mountain View',
 'cod': 200}

In [8]:
import json

def getSentiment(text):
    endpoint = "https://gateway.watsonplatform.net/natural-language-understanding/api/v1/analyze"
    # Must have own username and password for TBM Watson API
    username = " " 
    password =" "
    parameters = {
        'features': 'emotion,sentiment',
        'version' :  '2019-07-12',
        'text': text,
        'language' : 'en',
        # url = url_to_analyze, this is an alternative to sending the text
    }

    resp = requests.get(endpoint, params=parameters, auth=(username, password))
    return resp.json()


In [9]:
# We will analyze the text below using the IBM Watson API

text = '''
I say to you today, my friends, though, even though we face the difficulties of today and tomorrow, I still 
©2014 The Gilder Lehrman Institute of American History
www.gilderlehrman.org
have a dream. It is a dream deeply rooted in the American dream. I have a dream that one day this
nation will rise up, live out the true meaning of its creed: “We hold these truths to be self-evident, that
all men are created equal.”
I have a dream that one day on the red hills of Georgia sons of former slaves and the sons of former
slave-owners will be able to sit down together at the table of brotherhood. I have a dream that one day
even the state of Mississippi, a state sweltering with the heat of injustice, sweltering with the heat of
oppression, will be transformed into an oasis of freedom and justice.
I have a dream that my four little children will one day live in a nation where they will not be judged by
the color of their skin but by the content of their character. I have a dream . . . I have a dream that one
day in Alabama, with its vicious racists, with its governor having his lips dripping with the words of
interposition and nullification, one day right there in Alabama little black boys and black girls will be able
to join hands with little white boys and white girls as sisters and brothers.
I have a dream today . . .
This will be the day when all of God’s children will be able to sing with new meaning. “My country, ’tis of
thee, sweet land of liberty, of thee I sing. Land where my fathers died, land of the pilgrim’s pride, from
every mountain side, let freedom ring.” And if America is to be a great nation, this must become true. So
let freedom ring from the prodigious hilltops of New Hampshire. Let freedom ring from the mighty
mountains of New York. Let freedom ring from the heightening Alleghenies of Pennsylvania. Let
freedom ring from the snowcapped Rockies of Colorado. Let freedom ring from the curvaceous slopes of
California.
But not only that. Let freedom ring from Stone Mountain of Georgia. Let freedom ring from Lookout
Mountain of Tennessee. Let freedom ring from every hill and molehill of Mississippi, from every
mountain side. Let freedom ring . . .
When we allow freedom to ring—when we let it ring from every city and every hamlet, from every state
and every city, we will be able to speed up that day when all of God’s children, black men and white
men, Jews and Gentiles, Protestants and Catholics, will be able to join hands and sing in the words of the
old Negro spiritual, “Free at last, Free at last, Great God a-mighty, We are free at last.”
'''

In [10]:
data = getSentiment(text)
data

{'usage': {'text_units': 1, 'text_characters': 2594, 'features': 2},
 'sentiment': {'document': {'score': 0.731566, 'label': 'positive'}},
 'language': 'en',
 'emotion': {'document': {'emotion': {'sadness': 0.4218,
    'joy': 0.59039,
    'fear': 0.083406,
    'disgust': 0.048542,
    'anger': 0.103113}}}}

In [11]:
data['sentiment']['document']

{'score': 0.731566, 'label': 'positive'}

In [12]:
data['emotion']['document']['emotion']

{'sadness': 0.4218,
 'joy': 0.59039,
 'fear': 0.083406,
 'disgust': 0.048542,
 'anger': 0.103113}

This text is an excerpt taken from Martin Luther King's famous I Have a Dream speech, which I assumed would be largely positive and hopeful. This is somewhat shown in the analysis because the sentiment that was given to the speech is positive and because the score given is 0.731566 that means the speech is very strongly positive. This can also be seen from the emotion detected in the speech where the joy is given a 0.59039. One thing I didn't expect is there to be as high of a score given to sadness however, I guess it could be coming from the many times he talks about the hardships African Americans have faced and the joy is coming from the hope he is trying to inspire in the face of these hardships.

In [13]:
def processURL(url_to_analyze):
    endpoint_watson = "https://gateway.watsonplatform.net/natural-language-understanding/api/v1/analyze"
    params = {
        'version': '2019-07-12',
    }
    headers = { 
        'Content-Type': 'application/json',
    }
    watson_options = {
      "url": url_to_analyze,
      "features": {
        "entities": {
          "sentiment": True,
          "emotion": True,
          "limit": 10
        }
      }
    }
    username = "apikey"
    password = "nZChfYjcoh8-dMED4bA_dkeoOVZsLwhoCVw-zw2g5TcM"

    resp = requests.post(endpoint_watson, 
                         data=json.dumps(watson_options), 
                         headers=headers, 
                         params=params, 
                         auth=(username, password) 
                        )
    return resp.json()

In [43]:
endpoint = "https://newsapi.org/v2/top-headlines?country=us&category=business&apiKey= " # Must have own apiKey
resp = requests.get(endpoint)
articles = resp.json()

In [45]:
url_to_analyze = articles['articles'][3]['url']

http://techcrunch.com/2020/02/10/tesla-ramps-up-solar-tile-roofs-installations-in-u-s-eyes-china-and-europe-expansion/


In [ ]:
data = processURL(url_to_analyze)
print(url_to_analyze)

In [46]:
data["entities"][0]

{'type': 'Company',
 'text': 'Tesla',
 'sentiment': {'score': 0.73207, 'label': 'positive'},
 'relevance': 0.964454,
 'emotion': {'sadness': 0.285716,
  'joy': 0.535247,
  'fear': 0.071767,
  'disgust': 0.049258,
  'anger': 0.046916},
 'disambiguation': {'subtype': ['AutomobileCompany', 'VentureFundedCompany'],
  'name': 'Tesla_Motors',
  'dbpedia_resource': 'http://dbpedia.org/resource/Tesla_Motors'},
 'count': 7,
 'confidence': 1}

In [47]:
def getEntities(data, threshold):
    result = [print('Entity: ' + entity['text'] + ' Relevance: ' + str(entity['relevance']) + ' Associated Sentiment: ' + 
            str(entity['sentiment']['score']))  for entity in data["entities"] if float(entity['relevance']) > threshold]
    return result

getEntities(data, 0.25)

Entity: Tesla Relevance: 0.964454 Associated Sentiment: 0.73207
Entity: Musk Relevance: 0.537669 Associated Sentiment: 0.64055
Entity: CEO Elon Musk Relevance: 0.528137 Associated Sentiment: 0.62164
Entity: Buffalo Relevance: 0.362029 Associated Sentiment: 0
Entity: San Francisco Bay Relevance: 0.290189 Associated Sentiment: 0.62164


[None, None, None, None, None]